content/courses/advanced-ada/parts/resource_management/controlled_types.rst

```diff
@@ -569,16 +569,513 @@ Finally, the new version of object :ada:`A` (the one whose ID is 3) and object
+
+    package Simple_Controlled_Types is
+
+       type T (S1 : access Selection;
+               S2 : access Selection) is
+         tagged private;
+
+       procedure Dummy (E : T);
+
+    private
+
+       function Default_Init return Integer;
+
+       type T (S1 : access Selection;
+               S2 : access Selection) is new
+         Ada.Finalization.Controlled with
+       record
+          Sel_1 : Selection_1 (S1);
+          Sel_2 : Selection_2 (S2);
+          S_1   : Sub_1;
+          I     : Integer := Default_Init;
+       end record;
+
+       overriding
+       procedure Initialize (E : in out T);
+
+       overriding
+       procedure Finalize (E : in out T);
+
+    end Simple_Controlled_Types;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    package body Simple_Controlled_Types is
+
+       function Default_Init return Integer is
+       begin
+          Put_Line ("Default_Init: Integer...");
+          return 42;
+       end Default_Init;
+
+       procedure Dummy (E : T) is
+       begin
+          Put_Line ("(Dummy: T...)");
+       end Dummy;
+
+       procedure Initialize (E : in out T) is
+       begin
+          Put_Line ("Initialize: T...");
+       end Initialize;
+
+       procedure Finalize (E : in out T) is
+       begin
+          Put_Line ("Finalize: T...");
+       end Finalize;
+
+    end Simple_Controlled_Types;
+
+    with Simple_Controlled_Types;
+    use  Simple_Controlled_Types;
+
+    with Selections;
+    use  Selections;
+
+    procedure Show_Controlled_Types is
+       S1, S2 : aliased Selection;
+       A : T (S1'Access, S2'Access);
+    begin
+       Dummy (A);
+    end Show_Controlled_Types;
+
+When running this example, we see that all other subcomponents |mdash| to be
+more precise, those subcomponents that require initialization |mdash| are
+initialized before the :ada:`Sub_1` and :ada:`Sub_2` components are initialized
+via calls to their corresponding :ada:`Initialize` procedure. Note that,
+although :ada:`Sub_1` and :ada:`Sub_2` are the last components to be
+initialized, they are still initialized before the call to the
+:ada:`Initialize` procedure of type :ada:`T`.
+
+Task activation
+~~~~~~~~~~~~~~~
+
+Components of task types also require special treatment. According to the Ada
+Reference Manual, "for an allocator, any task activations follow all calls on
+:ada:`Initialize`."
+
+As always, let's analyze an example that illustrates this. First, we implement
+another package called :ada:`Workers` with a simple task type:
+
+.. code:: ada compile_button project=Courses.Advanced_Ada.Resource_Management.Controlled_Types.Initialization.Controlled_Initialization
+
+    package Workers is
+
+       task type Worker is
+          entry Start;
+          entry Stop;
+       end Worker;
+
+    end Workers;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    package body Workers is
+
+       task body Worker is
+       begin
+          Put_Line ("Activating Worker task...");
```

Comment:
You might want to move this Put_Line call to a function called within the declarative part, since activation of a given task is the elaboration of the task object's declarative part. But that might be more precision than necessary...

Reply:
On the contrary: I think this is an excellent idea! I've just changed the code accordingly. Thanks for the suggestion! 

Also, I took the opportunity to add some diagrams with the calling sequence.